lib/elixir/lib/partition_supervisor.ex

```diff
@@ -0,0 +1,315 @@
+  become a bottleneck. To address this, you can start multiple instances
+  of the `DynamicSupervisor` and then pick a "random" instance to start
+  the child on.
+
+  Instead of:
+
+      children = [
+        {DynamicSupervisor, name: MyApp.DynamicSupervisor}
+      ]
+
+      Supervisor.start_link(children, strategy: :one_for_one)
+
+  and:
+
+      DynamicSupervisor.start_child(MyApp.DynamicSupervisor, {Agent, fn -> %{} end})
+
+  You can do this:
+
+      children = [
+        {PartitionSupervisor,
+         child_spec: DynamicSupervisor,
+         name: MyApp.DynamicSupervisors}
+      ]
+
+      Supervisor.start_link(children, strategy: :one_for_one)
+
+  and then:
+
+      DynamicSupervisor.start_child(
+        {:via, PartitionSupervisor, {MyApp.DynamicSupervisors, self()}},
+        {Agent, fn -> %{} end}
+      )
+
+  In the code above, we start a partition supervisor that will by default
+  start a dynamic supervisor for each core in your machine. Then, instead
+  of calling the `DynamicSupervisor` by name, you call it through the
+  partition supervisor using the `{:via, PartitionSupervisor, {name, key}}`
+  format. We picked `self()` as the routing key, which means each process
+  will be assigned one of the existing dynamic supervisors. See `start_link/1`
+  to see all options supported by the `PartitionSupervisor`.
+
+  ## Implementation notes
+
+  The `PartitionSupervisor` requires a name as an atom to be given on start,
+  as it uses an ETS table to keep all of the partitions. Under the hood,
+  the `PartitionSupervisor` generates a child spec for each partition and
+  then act as a regular supervisor. The id of each child spec is the
+  partition number.
+
+  For routing, two strategies are used. If `key` is an integer, it is routed
+  using `rem(abs(key), partitions)`. Otherwise it uses `:erlang.phash2(key, partitions)`.
+  The particular routing may change in the future, and therefore cannot
+  be relied on. If you want to retrieve a particular PID for a certain key,
+  you can use `GenServer.whereis({:via, PartitionSupervisor, {name, key}})`.
+  """
+
+  @behaviour Supervisor
+  @type name :: atom()
+
+  @doc false
+  def child_spec(opts) when is_list(opts) do
+    %{
+      id: Keyword.get(opts, :name, PartitionSupervisor),
+      start: {PartitionSupervisor, :start_link, [opts]},
+      type: :supervisor
+    }
+  end
+
+  @doc """
+  Starts a partition supervisor with the given options.
+
+  This function is typically not invoked directly, instead it is invoked
+  when using a `PartitionSupervisor` as a child of another supervisor:
+
+      children = [
+        {PartitionSupervisor, child_spec: SomeChild, name: MyPartitionSupervisor}
+      ]
+
+  If the supervisor is successfully spawned, this function returns
+  `{:ok, pid}`, where `pid` is the PID of the supervisor. If the supervisor
+  is given a name and a process with the specified name already exists,
```

Comment:
Minor nitpick: "if the supervisor is given a name" sounds like the name can be optionally given, but it is mandatory

```suggestion
  `{:ok, pid}`, where `pid` is the PID of the supervisor. If the given name
    for the partition supervisor is already assigned to a process,
```